src/Ui/Table.elm

```diff
@@ -0,0 +1,464 @@
+module Ui.Table exposing
+    ( ColumnAlignment(..)
+    , ColumnConfig
+    , Flex(..)
+    , Model
+    , Msg
+    , SortDirection(..)
+    , TableConfig
+    , addColumn
+    , addExtendableView
+    , cellColumns
+    , cellCustom
+    , cellFloat
+    , cellInt
+    , cellLink
+    , cellRows
+    , cellText
+    , column
+    , columnWithDataEntryId
+    , defaultConfig
+    , hideColumn
+    , setColumnAlignment
+    , setColumnWidth
+    , showColumn
+    , update
+    , view
+    )
+
+import Css
+import Css.Transitions exposing (transition)
+import Dict exposing (Dict)
+import Html.Styled exposing (Attribute, Html, a, div, span)
+import Html.Styled.Attributes exposing (css, href)
+import Html.Styled.Events exposing (onClick)
+import Html.Styled.Keyed exposing (lazyNode3)
+import List
+import Maybe.Extra exposing (isJust)
+import Ui.Css.Palette as Palette
+import Ui.Css.TextStyle exposing (toCssStyle)
+import Ui.Icon
+import Ui.Palette
+import Ui.Text as Text
+import Ui.TextStyle as TextStyle exposing (TextStyle)
+
+
+type SortDirection
+    = Ascending
+    | Descending
+
+
+type ColumnAlignment
+    = Left
+    | Center
+    | Right
+
+
+type ColumnContent msg
+    = ColumnContent (Html msg)
+
+
+type alias ColumnConfig record msg =
+    { name : String
+    , width : Css.LengthOrAuto Css.Px
+    , content : DataEntryId -> record -> ColumnContent msg
+    , visible : Bool
+    , alignment : ColumnAlignment
+    , weight : Int
+    }
+
+
+type alias TableConfig record msg =
+    { columns : Dict ColumnId (ColumnConfig record msg)
+    , onHeaderClick : Maybe (ColumnId -> msg)
+    , extendableView : Maybe (record -> Html msg)
+    , liftMsg : Msg -> msg
+    }
+
+
+type alias Model record =
+    { sortDirection : SortDirection
+    , sortIndex : String
+    , data : Dict DataEntryId ( record, Bool )
+    }
+
+
+type alias ColumnId =
+    String
+
+
+type alias DataEntryId =
+    String
+
+
+type Msg
+    = ToggleAccordionView DataEntryId
```

Comment:
is this an extra view you can unfold under a row? Is this meant for the ingredient search? Are we sure this is useful to us?

Reply:
I'm pretty sure something like this will be useful in some tables such as the canvas view. Right now we have a lot of information on the display that isn't that interesting to compare but might be useful to see without having to navigate to another view. This is an experiment so we will see were we end up, maybe we will drop this in the end if we see no use for it! I have an accompanying exploration branch in the FE called explore-table-overhaul where I'm trying out different ways of displaying information using this component. It needs some manual wiring though since I haven't published this version of the package yet.